Implemented the app for slicing text given start and end time

bohJiang12 opened the pull request:
## Implementations

1. Changed the version of clams from 1.2.3 to 1.2.4
2. Added user/client input parameters for the app
3. Implemented necessary functions: `_metadata` and `_annotate`

## Results

The sliced text results are tested via:

1. Verify if additional `view` is added into the input `.mmif` file
2. Compare the sliced text is the same as the one returned by `text_document_helper.slice_text` of `clams-python=1.2.4`

## Questions

- Should the `TextDocument` be one of the inputs in app's metadata rather than only `TimeFrame` (while using `add_input` function)?

bohJiang12 committed:
Implemented the app for slicing text given start and end time

bohJiang12 committed:
Re-implemented app that adapts need for slicing text based on time frames

keighrim commented on requirements.txt:
```diff
@@ -1,2 +1,2 @@
 # Make sure clams-python version is explicitly specified, at least the lower bound
-clams-python==1.2.3
+clams-python==1.2.4
```
now 1.2.5 is release, let's bump this up as well. (check `Containerfile` as well. )

keighrim replied:
you need to update the base image version in containerfile as well. Also, there was a small bug in the SDK version that you used to generate the app template. line 28 of the Containerfile, 
https://github.com/clamsproject/app-text-slicer/blob/6253ad3c572e1116e029201a65178a023746efea/Containerfile#L28
the `&&` should be just `;`. So the line should be 
```
RUN mkdir /cache ; rm -rf /root/.cache ; ln -s /cache /root/.cache
```

keighrim commented on metadata.py:
```diff
@@ -40,13 +40,16 @@ def appmetadata() -> AppMetadata:
+    metadata.add_output(AnnotationTypes.Alignment)
+    
     
     # (optional) and finally add runtime parameter specifications
```
I think having a parameter to "select" TimeFrame's label would be a good idea. For example, a user can select "credits"-label timeframes only to slice text. Or multiple labels (e.g., chyrons & credit).

keighrim replied:
(probably by an untold convention) we use camelCase for parameter names instead of snake_case. Please rename the parameter name.

keighrim commented on app.py:
```diff
@@ -37,20 +39,36 @@ def _appmetadata(self):
+
+        return self.mmif
+
+        # for tf_view in mmif.get_all_views_contain(AnnotationTypes.TimeFrame):
```
I don't think we need to keep these old commented out lines. Let's get rid of these for readability.

bohJiang12 committed:
Added an option for users to select TimeFrame of wanted labels to slice

bohJiang12 commented:
## Fixes
1. Updated version of `clams-python` to `1.2.5`
2. Removed auxiliary comment blocks
3. Added an option/argument for the app, so that users are able to select `TimeFrame`s containing wanted labels, for example, "slate", "credits", "bars", etc.

## Feedback
* Using the `.mmif` file exported by the `whisper-wrapper` that I run at the very first time, an edge case for `TimeFrame` was its `start` and `end` properties were the same, therefore an assertion error would be raised since our `get_annotations_between_time` feature assumed `start < end`. So probably we need to fix this assertion to tolerate this edge case so that `start <= end`?
* In terms of input format that users select their wanted labels contained in `TimeFrame`, I implemented its type as `str`. Then, it's fine for users choose only one label, that is, `'credits'` for example. However, if multiple labels are selected, like `"credits, slate, chyron"`, I need to tokenized this parameter into a list of `str` as `['credits', 'slate', 'chyron']`. Since my initial and ideal thought towards this input format is to ask users to give a list/array, however, it seems that the app has no such type option. Do you think the `str` type input is enough for the app?

keighrim commented on app.py:
```diff
@@ -25,34 +26,37 @@
+        self.text_doc = self.mmif.get_documents_by_type(DocumentTypes.TextDocument)
+        assert len(self.text_doc) == 1, "There should be exactly one TextDocument in the MMIF file"
+
+        labels = parameters["contain_labels"]
```
if you set the parameter to be `multivalue=True' (https://sdk.clams.ai/appmetadata.html#runtime-parameter-specification) you don't need to do this "casting" yourself.

keighrim replied:
(in that case, the default should be `[]` instead of `*`.

keighrim commented on app.py:
```diff
@@ -25,34 +26,37 @@
+        for tf_view in self.mmif.get_all_views_contain(AnnotationTypes.TimeFrame):
+            tf_anns_in_view = tf_view.get_annotations(AnnotationTypes.TimeFrame)
+            for tf_ann in tf_anns_in_view:
+                if not label_set or tf_ann._get_label() in label_set:
```
where did find this `_get_label()` :rofl: I think you can just use `get_property('label')` as the `_get_label()` is an obsolete method. (which now I think I should put a deprecation warning there).

bohJiang12 replied:
To be honest, I won't trust copilot anymore. I intended to use `get_property` until it suggested `_get_label()` instead.

keighrim replied:
Yeah, it's there because of a historical reason. https://github.com/clamsproject/mmif/issues/218 So copilot isn't entirely wrong, I guess :melting_face:

keighrim commented on app.py:
```diff
@@ -25,34 +26,37 @@
+                    end_time = self.mmif.get_end(tf_ann) 
+                    sliced_text = new_view.new_textdocument(text_document_helper.slice_text(self.mmif, start_time, end_time))
+                    new_align = new_view.new_annotation(at_type=AnnotationTypes.Alignment,
+                                                        properties={'source': tf_ann.id, 'target': sliced_text.id}) 
```
Just for safety, let's use `long_id` to make the identifiers fully unambiguous.

keighrim commented on metadata.py:
```diff
@@ -40,15 +40,27 @@ def appmetadata() -> AppMetadata:
+                           'lower-third', 'credits']
+                           User needs to provide a string of comma-separated labels, 
+                           e.g. 'bars, credits, chyron'.
+                           Default is all are selected.
```
If you really want to give a predefine choices, you can use `choices=['x', 'y', 'z']` argument. But in this case, since we can come up with a different name of the frame label in the future, I don't think it's safe to limit the filter names here.

keighrim reviewed (changes requested):
requesting more not-so-significant changes, I think this is almost done.

bohJiang12 committed:
Updated clams version to use and corrected RUN command

bohJiang12 committed:
Enabled multivalued parameter and polish some code usages

bohJiang12 commented:
This is the updated app implementations that 
* Updated docker base image using `clams-python=1.2.5` in `Containerfile`
* Enabled `multivalued=True` for the parameter `containLabels`
* Replaced unused `_get_label` by `get_property`
* Accessed `Annotation`'s id using `long_id` property

keighrim commented on metadata.py:
```diff
@@ -40,27 +40,27 @@ def appmetadata() -> AppMetadata:
-                           User needs to provide a string of comma-separated labels, 
-                           e.g. 'bars, credits, chyron'.
-                           Default is all are selected.
+                           User needs to provide a list of labels if multiple labels are expected
```
This is not really true. For multivalued parameters, users need to do something like [this](https://github.com/clamsproject/clams-python/issues/122#issuecomment-1513424021) . (SDK will make sure the `_annotate` method receives a list from that URL string format)

keighrim replied:
And, this URL format should be something users need to know first (or learn from the general app manual), so that doesn't need to be fully spelled out here, I guess. Let's just take out the format-related explanation.

bohJiang12 replied:
> For multivalued parameters, users need to do something like this

Does this mean that we expect users to run our app as `curl -X ... localhost:xxxx?containLabels=credits&containLabels=slate` if they're interested in `TimeFrame`s that contain `label` property "credits" and "slate"?

keighrim replied:
Yes, exactly.

keighrim reviewed (changes requested):
That small fix will end up in v1.0 !

bohJiang12 committed:
Completed README.md and edited help message in metadata.py

keighrim commented on app.py:
```diff
@@ -39,15 +39,16 @@ def _annotate(self, mmif: Mmif, **parameters) -> Mmif:
         self.text_doc = self.mmif.get_documents_by_type(DocumentTypes.TextDocument)
         assert len(self.text_doc) == 1, "There should be exactly one TextDocument in the MMIF file"
+        label_set = set(parameters["containLabel"])
+        assert len(label_set) >= 1, "There should be at least one label in the parameters"  # Here we make the label parameter as required
```
SDK will do check this when nothing passed for a "required" parameter. To make a parameter required, just take out `default=` argument when setting a parameter up. 

*(SDK will do the check, but due to this bug https://github.com/clamsproject/clams-python/issues/232 , the output will be messy for now. )

keighrim commented on metadata.py:
```diff
@@ -40,15 +40,26 @@ def appmetadata() -> AppMetadata:
-    # CHANGE this line and make sure return the compiled `metadata` instance
+    metadata.add_parameter(name='containLabel',
+                           description=
+                           '''
```
this will end up like this 
``` json 
 $ curl localhost:5000 | jq 
{
  ...
  "parameters": [
    {
      "name": "containLabel",
      "description": "\n                           A list of labels that user expect TimeFrames contain.\n                           Available options are:\n                           ['bars', 'tones', 'bars-and-tones','speech',\n                           'noise', 'music', 'slate', 'chyron',\n                           'lower-third', 'credits']\n                           Users must select at least one of the following labels, errors would \n                           be thrown instead.\n                           ",
      ...
    },
    ...
  ]
  ...
}
```
When using three-quotes string, be aware of the difference btw beautiful code and actual output. 

Here, I think what you really want is not a "multi-line" string ([pep-295](https://peps.python.org/pep-0295/), rejected) but a string value in continuation lines ([pep-8](https://peps.python.org/pep-0008/), also [string literal in python parser reference](https://docs.python.org/3/reference/lexical_analysis.html#string-literal-concatenation).

keighrim replied:
Also, "Available options" are really limited to those items, so it'd better to say that the list is "non-exhaustive", or "not limited to", or change the wording from "Available options" to "For example" or something like that.

bohJiang12 replied:
>When using three-quotes string, be aware of the difference btw beautiful code and actual output.

I agree. Since I tested the app description in CLI by calling `python cli.py -h`, but the description hasn't beed divided into multi-lines if I add `\n` in the string. 

Now I changed the string format of this argument as 
```
 add_parameter(name=....
               description=".....\n"
                           ".....\n"
                           ".....\n"
               other_params=...
               )
```
So that it makes both code and actual output pretty. What's your thought?

keighrim reviewed (changes requested):
small changes!

bohJiang12 committed:
Polished string format and wording of parameter description
